calico-vpp-agent/prometheus/prometheus.go

```diff
@@ -0,0 +1,200 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus
+
+import (
+	"context"
+	"log"
+	"net/http"
+	"time"
+
+	"strconv"
+	"strings"
+	"sync"
+
+	"git.fd.io/govpp.git/adapter"
+	"git.fd.io/govpp.git/adapter/statsclient"
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"github.com/orijtech/prometheus-go-metrics-exporter"
+	"github.com/projectcalico/vpp-dataplane/calico-vpp-agent/cni/storage"
+	"github.com/projectcalico/vpp-dataplane/vpplink"
+	"github.com/sirupsen/logrus"
+)
+
+type Event int
+
+const (
+	Add    Event = 0
+	Delete Event = 1
+
+	recordMetricInterval int64 = 5
+)
+
+type Server struct {
+	log                      *logrus.Entry
+	vpp                      *vpplink.VppLink
+	podInterfacesBySwifIndex map[uint32]storage.LocalPodSpec
+	podInterfacesByKey       map[string]storage.LocalPodSpec
+	sc                       *statsclient.StatsClient
+	channel                  chan PodSpecEvent
+	lock                     sync.Mutex
+}
+
+type PodSpecEvent struct {
+	Event Event
+	Pod   storage.LocalPodSpec
+}
+
+func (s *Server) recordMetrics() {
+	go func() {
+		pe, err := prometheus.New(prometheus.Options{})
+		if err != nil {
+			log.Fatalf("Failed to create new exporter: %v", err)
+		}
+		mux := http.NewServeMux()
+		mux.Handle("/metrics", pe)
+		go func() {
+			http.ListenAndServe(":8888", mux)
+		}()
+		for {
+			time.Sleep(time.Second * time.Duration(recordMetricInterval))
+			ifNames, dumpStats, _ := vpplink.StatsClientFunc(s.sc)
+			for _, sta := range dumpStats {
+				if string(sta.Name) != "/if/names" {
+					names := []string{strings.Replace(string(sta.Name[4:]), "-", "_", -1)}
+					if sta.Type == adapter.CombinedCounterVector {
+						names = []string{names[0] + "_packet", names[0] + "_bytes"}
+					}
+					for k, name := range names {
+						metric := &metricspb.Metric{
+							MetricDescriptor: &metricspb.MetricDescriptor{
+								Name:        name,
+								Description: "Extra ones",
+								Unit:        "1",
+								LabelKeys: []*metricspb.LabelKey{
+									{Key: "worker", Description: "worker"},
+									{Key: "namespace", Description: "Operating system"},
+									{Key: "podName", Description: "Name of the pod"},
+									{Key: "nameInPod", Description: "Name pf interface in the pod"},
+								},
+							},
+							Timeseries: []*metricspb.TimeSeries{},
+						}
+						s.lock.Lock()
+						if sta.Type == adapter.SimpleCounterVector {
+							values := sta.Data.(adapter.SimpleCounterStat)
+							for worker := range values {
+								for ifIdx := range values[worker] {
+									if string(ifNames[ifIdx]) != "" {
+										_, swifindex := s.vpp.SearchInterfaceWithName(string(ifNames[ifIdx]))
```

Comment:
It's not going to be acceptable to do this many API calls to VPP just to gather stats, it would have too much impact on the dataplane, especially since SearchInterfaceWithName is not optimized.
I think ifIdx should already be the sw_if_index, so there should be no need to query vpp.